pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -46,19 +46,91 @@ const (
 	// RbacConfigName is the name of the RbacConfig custom resource that controls the RBAC behavior.
 	RbacConfigName = "rbac-config"
 
-	sourceIP        = "source.ip"
-	sourceService   = "source.service"
-	destinationIP   = "destination.ip"
-	destinationPort = "destination.port"
-	userHeader      = ":user"
-	methodHeader    = ":method"
-	pathHeader      = ":path"
-	//TODO(yangminzhu): Update the key after the final header name of service is decided.
-	serviceHeader       = ":service"
-	requestHeaderPrefix = "request.header["
-	requestHeaderSuffix = "]"
+	// attributes that could be used in both ServiceRole and ServiceRoleBinding.
+	attributeIP     = "ip"
+	attributeHeader = "header:" // header name is after the colon, e.g. "header:User-Agent".
+
+	// attributes that could be used in a ServiceRole constraint.
+	attributePort      = "port"
+	attributeLabel     = "label:"          // label name is after the colon, e.g. "label:version".
+	attributeName      = "name"            // short service name, e.g. "productpage".
+	attributeNamespace = "namespace"       // e.g. "default".
+	attributeSA        = "service-account" // e.g. "bookinfo-productpage"
+
+	// TODO(yangminzhu): Add authn related attributes.
+
+	userHeader   = ":user"
+	methodHeader = ":method"
+	pathHeader   = ":path"
 )
 
+// serviceMetadata is a collection of different kind of information about a service.
+type serviceMetadata struct {
+	name       string            // full qualified service name.
+	labels     map[string]string // labels of the service instance
+	attributes map[string]string // additional attributes of the service
+}
+
+func createServiceMetadata(attr *model.ServiceAttributes, in *model.ServiceInstance) *serviceMetadata {
+	return &serviceMetadata{
+		name:   in.Service.Hostname.String(),
+		labels: in.Labels,
+		attributes: map[string]string{
+			attributeName:      attr.Name,
+			attributeNamespace: attr.Namespace,
+			attributeSA:        in.ServiceAccount,
+		},
+	}
+}
+
+// shouldMatchAttribute returns true if the given attribute should be matched just in the plugin.
+// This is because we already have enough information to finish this in the plugin.
+func shouldMatchAttribute(attr string) bool {
+	switch attr {
+	case attributeName, attributeNamespace, attributeSA:
```

Comment:
should you pass in "attributes" in serviceMetadata to this function? How do you compare attribute name like "destination.service" to constants like "name"?

Reply:
we do not have any attributes concept in pilot, so it actually doesn't matter if we choose "destination.name" or "name". It's just used by the user to specify what value they're expecting. In the runtime, we extract and compare it to multiple sources (like `model.ServiceInstance`)

User create the ServiceRole like this:
```
Service: ["productpage.default.svc.local"]
Method: ["GET"]
Constraints:
  - "name": ["productpage"]
  - "label:version": ["v1", "v2"]
```
Or like this (if choose to use long name):
```
Service: ["productpage.default.svc.local"]
Method: ["GET"]
Constraints:
  - "destination.name": ["productpage"]
  - "destination.label:version": ["v1", "v2"]
```

Because the constraint part is already explicit and obvious that it's for the destination, so I feel a shorter name is more easier to write.

Reply:
Note: I also changed the format of label (and header) from `label[version]` to `label:version`, This is more efficient to parse and easy to write.

Reply:
I understand there is no attribute concept in Pilot. However, "attribute" is Istio concept. Istio users have good understanding of attributes, and they are the standard way to describe the requests and environments. In our local RBAC proposal, we proposed to let users use attributes directly to describe the constraints and principal properties. The examples you gave above should be written as follows:
```
  services: ["productpage.default.svc.local"]
  methods: ["GET"]
  constraints:
  - key: destination.name
    values: ["productpage"]
  - key: destination.labels["version"]
    values: ["v1", "v2"]
```

Reply:
Done